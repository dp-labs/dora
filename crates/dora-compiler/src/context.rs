use melior::{
    Context as MLIRContext,
    dialect::DialectRegistry,
    utility::{register_all_dialects, register_all_llvm_translations, register_all_passes},
};
use std::path::PathBuf;

/// The `Context` struct represents the MLIR (Multi-Level Intermediate Representation)
/// context in which various operations, types, and other IR constructs exist.
/// It serves as the core environment for creating, managing, and manipulating
/// MLIR-based intermediate representations.
///
/// # Fields:
/// - `mlir_context`: The `MLIRContext` is the underlying context used to manage
///   all MLIR-related constructs, ensuring proper memory management and scoping
///   of IR elements.
///
/// # Example Usage:
/// ```no_check
/// let ctx = Context::new();
/// ```
///
/// This struct is essential when working with MLIR in a Rust-based environment,
/// providing the foundational context to operate on IR.
#[derive(Debug, Eq, PartialEq)]
pub struct Context {
    pub mlir_context: MLIRContext,
}

unsafe impl Send for Context {}
unsafe impl Sync for Context {}

impl Default for Context {
    fn default() -> Self {
        Self::new()
    }
}

impl Context {
    pub fn new() -> Self {
        let mlir_context = initialize_mlir();
        Self { mlir_context }
    }
}

/// The `Session` struct represents a session in which an MLIR file is processed.
/// It contains paths to the raw and after-pass MLIR files, allowing the user to track
/// different stages of the MLIR transformation process.
///
/// # Fields:
/// - `raw_mlir_path`: This field holds an optional `PathBuf` pointing to the path
///   of the raw MLIR file before any transformations or passes have been applied.
///   It is `None` if the raw MLIR file path is not provided.
/// - `after_pass_mlir_path`: This field holds an optional `PathBuf` pointing to the path
///   of the MLIR file after various passes and transformations have been applied.
///   It is `None` if the after-pass MLIR file path is not provided.
///
/// # Example Usage:
/// ```no_check
/// let session = Session {
///     raw_mlir_path: Some(PathBuf::from("input.mlir")),
///     after_pass_mlir_path: Some(PathBuf::from("output.mlir")),
/// };
/// ```
///
/// The `Session` struct helps to keep track of MLIR files at different stages
/// during compilation, especially in workflows where passes are applied to modify
/// the IR in successive steps.
#[derive(Debug, Clone, Default)]
pub struct Session {
    /// The path for the raw mlir file before any transformation passes.
    pub raw_mlir_path: Option<PathBuf>,
    /// The path for the MLIR file after transformation passes have been applied.
    pub after_pass_mlir_path: Option<PathBuf>,
}

/// Initializes and configures a new `MLIRContext` with all available dialects and passes.
///
/// This function sets up an `MLIRContext` by performing the following steps:
/// - Creates a new `MLIRContext`.
/// - Appends a registry of all available dialects.
/// - Attaches a diagnostic handler that logs diagnostics to `stderr`.
/// - Loads all available dialects into the context.
/// - Allows the use of unregistered dialects within the context.
/// - Registers all available optimization and transformation passes.
/// - Registers LLVM translations for further compilation steps.
///
/// # Returns
///
/// A fully initialized `MLIRContext` ready for compiling and optimizing MLIR-based operations.
///
/// # Behavior
///
/// - This context is configured to handle diagnostics and allows unregistered dialects to be used.
/// - It includes all MLIR dialects and passes available in the current setup.
/// - The context supports LLVM translations for lowering MLIR to LLVM IR.
pub fn initialize_mlir() -> MLIRContext {
    let context = MLIRContext::new();
    context.append_dialect_registry(&{
        let registry = DialectRegistry::new();
        register_all_dialects(&registry);
        registry
    });
    context.attach_diagnostic_handler(|diagnostic| {
        eprintln!("{diagnostic}");
        true
    });
    context.load_all_available_dialects();
    context.set_allow_unregistered_dialects(true);
    register_all_passes();
    register_all_llvm_translations(&context);
    context
}
