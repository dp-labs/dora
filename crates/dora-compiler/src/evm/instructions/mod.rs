use melior::ir::BlockRef;

use crate::conversion::builder::OpBuilder;
use crate::evm::EVMCompiler;
use crate::evm::backend::EVMBuilder;

use super::CtxType;

pub mod arithmetic;
pub mod bitwise;
pub mod contract;
pub mod control;
pub mod host;
pub mod host_env;
pub mod memory;
pub mod stack;
pub mod system;

impl<'c> EVMCompiler<'c> {
    #[inline]
    pub(crate) fn make_builder<'a>(
        ctx: &'a mut CtxType<'c>,
        block: BlockRef<'c, 'a>,
    ) -> EVMBuilder<'a, 'c> {
        EVMBuilder {
            builder: OpBuilder::new_with_block(ctx.context, block),
            ctx,
            use_static_stack: false,
        }
    }
}
