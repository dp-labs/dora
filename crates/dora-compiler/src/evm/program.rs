use dora_primitives::spec::SpecId;
use num_bigint::BigUint;
pub use revmc::{op_info_map, OpcodeInfo};
use std::fmt;
use thiserror::Error;

/// An error that occurs when an invalid opcode is encountered during parsing.
/// This struct holds the invalid opcode (as a `u8`) and provides a formatted error message
/// indicating the invalid opcode in hexadecimal format.
///
/// # Example Usage:
/// ```no_check
/// let err = OpcodeParseError(0xFF);
/// println!("{}", err); // Output: "The opcode `FF` is not valid"
/// ```
///
/// # Notes:
/// - This error is triggered when the byte sequence does not match any valid opcode during
///   the parsing process, which is common when processing raw EVM bytecode.
#[derive(Error, Debug)]
#[error("The opcode `{:02X}` is not valid", self.0)]
pub struct OpcodeParseError(u8);

/// An error type that aggregates multiple `OpcodeParseError` instances. This is used when
/// multiple invalid opcodes are encountered during the parsing of bytecode.
///
/// # Fields:
/// - `0`: A vector of `OpcodeParseError` instances, each representing an individual invalid opcode.
///
/// # Example Usage:
/// ```no_check
/// let errors = vec![OpcodeParseError(0x01), OpcodeParseError(0xFF)];
/// let parse_error = ParseError(errors);
/// println!("{:?}", parse_error); // Output: ParseError([OpcodeParseError(0x01), OpcodeParseError(0xFF)])
/// ```
///
/// # Notes:
/// - This error type allows for handling cases where multiple parsing errors need to be reported,
///   such as when processing large sections of bytecode that contain multiple invalid opcodes.
#[derive(Error, Debug)]
pub struct ParseError(Vec<OpcodeParseError>);

impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let opcodes: Vec<_> = self.0.iter().map(|x| format!("{:02X}", x.0)).collect();
        writeln!(f, "The following opcodes could not be parsed: ")?;
        writeln!(f, "{:#?}", opcodes)?;
        Ok(())
    }
}

macro_rules! opcodes {
    ($($name:ident = $value:expr),+ $(,)?) => {
        #[repr(u8)]
        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
        pub enum Opcode {
            $($name = $value),+
        }

        impl TryFrom<u8> for Opcode {
            type Error = OpcodeParseError;

            fn try_from(opcode: u8) -> Result<Self, Self::Error> {
                match opcode {
                    $(x if x == Opcode::$name as u8 => Ok(Opcode::$name),)+
                    _ => Err(OpcodeParseError(opcode))
                }
            }
        }
    };
}

opcodes! {
    STOP = 0x00,
    ADD = 0x01,
    MUL = 0x02,
    SUB = 0x03,
    DIV = 0x04,
    SDIV = 0x05,
    MOD = 0x06,
    SMOD = 0x07,
    ADDMOD = 0x08,
    MULMOD = 0x09,
    EXP = 0x0A,
    SIGNEXTEND = 0x0B,
    // unused 0x0C-0x0F
    LT = 0x10,
    GT = 0x11,
    SLT = 0x12,
    SGT = 0x13,
    EQ = 0x14,
    ISZERO = 0x15,
    AND = 0x16,
    OR = 0x17,
    XOR = 0x18,
    NOT = 0x19,
    BYTE = 0x1A,
    SHL = 0x1B,
    SHR = 0x1C,
    SAR = 0x1D,
    // unused 0x1E-0x1F
    KECCAK256 = 0x20,
    // unused 0x21-0x2F
    ADDRESS = 0x30,
    BALANCE = 0x31,
    ORIGIN = 0x32,
    CALLER = 0x33,
    CALLVALUE = 0x34,
    CALLDATALOAD = 0x35,
    CALLDATASIZE = 0x36,
    CALLDATACOPY = 0x37,
    CODESIZE = 0x38,
    CODECOPY = 0x39,
    GASPRICE = 0x3A,
    EXTCODESIZE = 0x3B,
    EXTCODECOPY = 0x3C,
    RETURNDATASIZE = 0x3D,
    RETURNDATACOPY = 0x3E,
    EXTCODEHASH = 0x3F,
    BLOCKHASH = 0x40,
    COINBASE = 0x41,
    TIMESTAMP = 0x42,
    NUMBER = 0x43,
    PREVRANDAO = 0x44,
    GASLIMIT = 0x45,
    CHAINID = 0x46,
    SELFBALANCE = 0x47,
    BASEFEE = 0x48,
    BLOBHASH = 0x49,
    BLOBBASEFEE = 0x4A,
    // unused 0x4B-0x4F
    POP = 0x50,
    MLOAD = 0x51,
    MSTORE = 0x52,
    MSTORE8 = 0x53,
    SLOAD = 0x54,
    SSTORE = 0x55,
    JUMP = 0x56,
    JUMPI = 0x57,
    PC = 0x58,
    MSIZE = 0x59,
    GAS = 0x5A,
    JUMPDEST = 0x5B,
    TLOAD = 0x5C,
    TSTORE = 0x5D,
    MCOPY = 0x5E,
    PUSH0 = 0x5F,
    PUSH1 = 0x60,
    PUSH2 = 0x61,
    PUSH3 = 0x62,
    PUSH4 = 0x63,
    PUSH5 = 0x64,
    PUSH6 = 0x65,
    PUSH7 = 0x66,
    PUSH8 = 0x67,
    PUSH9 = 0x68,
    PUSH10 = 0x69,
    PUSH11 = 0x6A,
    PUSH12 = 0x6B,
    PUSH13 = 0x6C,
    PUSH14 = 0x6D,
    PUSH15 = 0x6E,
    PUSH16 = 0x6F,
    PUSH17 = 0x70,
    PUSH18 = 0x71,
    PUSH19 = 0x72,
    PUSH20 = 0x73,
    PUSH21 = 0x74,
    PUSH22 = 0x75,
    PUSH23 = 0x76,
    PUSH24 = 0x77,
    PUSH25 = 0x78,
    PUSH26 = 0x79,
    PUSH27 = 0x7A,
    PUSH28 = 0x7B,
    PUSH29 = 0x7C,
    PUSH30 = 0x7D,
    PUSH31 = 0x7E,
    PUSH32 = 0x7F,
    DUP1 = 0x80,
    DUP2 = 0x81,
    DUP3 = 0x82,
    DUP4 = 0x83,
    DUP5 = 0x84,
    DUP6 = 0x85,
    DUP7 = 0x86,
    DUP8 = 0x87,
    DUP9 = 0x88,
    DUP10 = 0x89,
    DUP11 = 0x8A,
    DUP12 = 0x8B,
    DUP13 = 0x8C,
    DUP14 = 0x8D,
    DUP15 = 0x8E,
    DUP16 = 0x8F,
    SWAP1 = 0x90,
    SWAP2 = 0x91,
    SWAP3 = 0x92,
    SWAP4 = 0x93,
    SWAP5 = 0x94,
    SWAP6 = 0x95,
    SWAP7 = 0x96,
    SWAP8 = 0x97,
    SWAP9 = 0x98,
    SWAP10 = 0x99,
    SWAP11 = 0x9A,
    SWAP12 = 0x9B,
    SWAP13 = 0x9C,
    SWAP14 = 0x9D,
    SWAP15 = 0x9E,
    SWAP16 = 0x9F,
    LOG0 = 0xA0,
    LOG1 = 0xA1,
    LOG2 = 0xA2,
    LOG3 = 0xA3,
    LOG4 = 0xA4,
    // unused 0xA5-0xEF
    CREATE = 0xF0,
    CALL = 0xF1,
    CALLCODE = 0xF2,
    RETURN = 0xF3,
    DELEGATECALL = 0xF4,
    CREATE2 = 0xF5,
    // unused 0xF6-0xF9
    RETURNDATALOAD = 0xF7,
    STATICCALL = 0xFA,
    // unused 0xFB-0xFC
    REVERT = 0xFD,
    INVALID = 0xFE,
    SELFDESTRUCT = 0xFF,
}

macro_rules! operations {
    ($(($variant:ident, $opcode:ident)),* $(,)?) => {
        #[derive(Debug, Clone)]
        pub enum Operation {
            $(
                $variant,
            )*
            PC { pc: usize },
            Jumpdest { pc: usize },
            Push((u8, BigUint)),
            Dup(u8),
            Swap(u8),
            Log(u8),
        }

        impl Operation {
            pub fn to(&self) -> Vec<u8> {
                match self {
                    $(
                        Operation::$variant => vec![Opcode::$opcode as u8],
                    )*
                    Operation::PC { .. } => vec![Opcode::PC as u8],
                    Operation::Jumpdest { .. } => vec![Opcode::JUMPDEST as u8],
                    Operation::Push((n, x)) => {
                        if !(1..=32).contains(n) {
                            panic!("Invalid PUSH size: {}", n);
                        }
                        let len = 1 + *n as usize;
                        let mut opcode_bytes = vec![0; len];
                        opcode_bytes[0] = Opcode::PUSH0 as u8 + n;
                        let bytes = x.to_bytes_be();
                        opcode_bytes[len - bytes.len()..].copy_from_slice(&bytes);
                        opcode_bytes
                    }
                    Operation::Dup(n) => {
                        if !(1..=16).contains(n) {
                            panic!("Invalid DUP index: {}", n);
                        }
                        vec![Opcode::DUP1 as u8 + n - 1]
                    },
                    Operation::Swap(n) => {
                        if !(1..=16).contains(n) {
                            panic!("Invalid SWAP index: {}", n);
                        }
                        vec![Opcode::SWAP1 as u8 + n - 1]
                    },
                    Operation::Log(n) => {
                        if !(0..=4).contains(n) {
                            panic!("Invalid LOG index: {}", n);
                        }
                        vec![Opcode::LOG0 as u8 + n]
                    }
                }
            }

            pub fn from(opcode: Opcode, opcodes: &[u8], pc: &mut usize) -> Result<Self, OpcodeParseError> {
                let op = match opcode {
                    $(
                        Opcode::$opcode => Operation::$variant,
                    )*
                    opcode if Opcode::PUSH1 as u8 <= opcode as u8 && opcode as u8 <= Opcode::PUSH32 as u8 => {
                            *pc += 1;
                            let n = (opcode as u8 - Opcode::PUSH0 as u8) as usize;
                            let x = opcodes[*pc..std::cmp::min(opcodes.len(), *pc + n)]
                                .try_into()
                                .unwrap();
                            *pc += n - 1;
                            Operation::Push((n as u8, BigUint::from_bytes_be(x)))
                    }
                    opcode if Opcode::DUP1 as u8 <= opcode as u8 && opcode as u8 <= Opcode::DUP16 as u8 => {
                        let index = (opcode as u8 - Opcode::DUP1 as u8 + 1) as u8;
                        Operation::Dup(index)
                    }
                    opcode if Opcode::SWAP1 as u8 <= opcode as u8 && opcode as u8 <= Opcode::SWAP16 as u8 => {
                        let index = (opcode as u8 - Opcode::SWAP1 as u8 + 1) as u8;
                        Operation::Swap(index)
                    }
                    opcode if Opcode::LOG0 as u8 <= opcode as u8 && opcode as u8 <= Opcode::LOG4 as u8 => {
                        let index = (opcode as u8 - Opcode::LOG0 as u8) as u8;
                        Operation::Log(index)
                    }
                    Opcode::PC => Operation::PC { pc: *pc },
                    Opcode::JUMPDEST => Operation::Jumpdest { pc: *pc },
                    _ => return Err(OpcodeParseError(opcode as u8)),
                };
                Ok(op)
            }
        }
    };
}

operations!(
    (Stop, STOP),
    (Add, ADD),
    (Mul, MUL),
    (Sub, SUB),
    (Div, DIV),
    (Sdiv, SDIV),
    (Mod, MOD),
    (SMod, SMOD),
    (Addmod, ADDMOD),
    (Mulmod, MULMOD),
    (Exp, EXP),
    (SignExtend, SIGNEXTEND),
    (Lt, LT),
    (Gt, GT),
    (Slt, SLT),
    (Sgt, SGT),
    (Eq, EQ),
    (IsZero, ISZERO),
    (And, AND),
    (Or, OR),
    (Xor, XOR),
    (Not, NOT),
    (Byte, BYTE),
    (Shl, SHL),
    (Shr, SHR),
    (Sar, SAR),
    (Keccak256, KECCAK256),
    (Address, ADDRESS),
    (Balance, BALANCE),
    (Origin, ORIGIN),
    (Caller, CALLER),
    (Callvalue, CALLVALUE),
    (CalldataLoad, CALLDATALOAD),
    (CallDataSize, CALLDATASIZE),
    (CallDataCopy, CALLDATACOPY),
    (Codesize, CODESIZE),
    (Codecopy, CODECOPY),
    (Gasprice, GASPRICE),
    (ExtcodeCopy, EXTCODECOPY),
    (ReturnDataSize, RETURNDATASIZE),
    (ReturnDataCopy, RETURNDATACOPY),
    (ReturnDataLoad, RETURNDATALOAD),
    (ExtcodeHash, EXTCODEHASH),
    (BlockHash, BLOCKHASH),
    (ExtcodeSize, EXTCODESIZE),
    (Coinbase, COINBASE),
    (Timestamp, TIMESTAMP),
    (Number, NUMBER),
    (Prevrandao, PREVRANDAO),
    (Gaslimit, GASLIMIT),
    (Chainid, CHAINID),
    (SelfBalance, SELFBALANCE),
    (Basefee, BASEFEE),
    (BlobHash, BLOBHASH),
    (BlobBaseFee, BLOBBASEFEE),
    (Pop, POP),
    (Mload, MLOAD),
    (Mstore, MSTORE),
    (Mstore8, MSTORE8),
    (Sload, SLOAD),
    (Sstore, SSTORE),
    (Jump, JUMP),
    (Jumpi, JUMPI),
    (Msize, MSIZE),
    (Gas, GAS),
    (Tload, TLOAD),
    (Tstore, TSTORE),
    (Mcopy, MCOPY),
    (Push0, PUSH0),
    (Create, CREATE),
    (Call, CALL),
    (CallCode, CALLCODE),
    (Return, RETURN),
    (DelegateCall, DELEGATECALL),
    (Create2, CREATE2),
    (StaticCall, STATICCALL),
    (Revert, REVERT),
    (Invalid, INVALID),
    (SelfDestruct, SELFDESTRUCT),
);

pub struct CompileOptions {
    pub spec_id: SpecId,
}

/// Represents a program that has been parsed and is ready for execution. The `Program` struct
/// holds a list of operations and the total code size of the bytecode it represents.
///
/// # Fields:
/// - `operations`: A vector of `Operation` structs, each representing an individual parsed operation
///   from the bytecode.
/// - `code_size`: The size of the bytecode, typically measured in bytes, representing the total length
///   of the code that was parsed.
///
/// # Example Usage:
/// ```no_check
/// use dora_compiler::evm::program::{Operation, Program};
/// use num_bigint::BigUint;
///
/// let program = Program {
///     operations: vec![
///         Operation::Push((1_u8, BigUint::from(1_u8))),
///         Operation::Push((1_u8, BigUint::from(1_u8))),
///         Operation::Add,
///     ],
///     code_size: 5,
/// };
/// println!("Program has {} operations and a code size of {} bytes", program.operations.len(), program.code_size);
/// ```
///
/// # Notes:
/// - The `Program` struct is used as the core representation of a parsed and compiled EVM bytecode.
///   It encapsulates the operations that will be executed by the EVM, along with metadata such as
///   the total size of the bytecode.
#[derive(Debug, Clone)]
pub struct Program {
    /// A vector of operations parsed from the bytecode.
    pub operations: Vec<Operation>,

    /// The total size of the bytecode (in bytes).
    pub code_size: u32,
}

impl Program {
    /// Constructs a `Program` from a slice of opcodes, checking for errors during parsing.
    ///
    /// This method attempts to parse the provided opcodes into operations and calculates
    /// the total code size. If any opcodes fail to parse, a `ParseError` is returned containing
    /// the failed opcodes.
    ///
    /// # Parameters
    /// * `opcodes` - A slice of bytes representing the opcodes to be parsed.
    ///
    /// # Returns
    /// - `Ok(Self)` - If all opcodes are successfully parsed.
    /// - `Err(ParseError)` - If any opcodes fail to parse, containing the list of failed opcodes.
    pub fn from_opcode_checked(opcodes: &[u8]) -> Result<Self, ParseError> {
        let (operations, failed_opcodes) = Self::parse_operations(opcodes);
        let code_size = Self::calculate_code_size(&operations);

        if failed_opcodes.is_empty() {
            Ok(Self {
                operations,
                code_size,
            })
        } else {
            Err(ParseError(failed_opcodes))
        }
    }

    /// Constructs a `Program` from a slice of opcodes without error checking.
    ///
    /// This method parses the provided opcodes into operations and calculates the total
    /// code size. It does not check for any parsing errors.
    ///
    /// # Parameters
    /// * `opcodes` - A slice of bytes representing the opcodes to be parsed.
    ///
    /// # Returns
    /// A `Program` instance constructed from the parsed operations.
    pub fn from_opcode(opcodes: &[u8]) -> Self {
        let (operations, _) = Self::parse_operations(opcodes);
        let code_size = Self::calculate_code_size(&operations);

        Self {
            operations,
            code_size,
        }
    }

    /// Converts the `Program` into a vector of opcodes.
    ///
    /// This method serializes the operations in the program back into their corresponding
    /// byte representations, returning a vector of opcodes.
    ///
    /// # Returns
    /// A vector of bytes representing the opcodes of the program.
    pub fn to_opcode(&self) -> Vec<u8> {
        self.operations
            .iter()
            .flat_map(Operation::to)
            .collect::<Vec<u8>>()
    }

    fn parse_operations(opcodes: &[u8]) -> (Vec<Operation>, Vec<OpcodeParseError>) {
        let mut operations = vec![];
        let mut failed_opcodes = vec![];
        let mut pc = 0;

        while pc < opcodes.len() {
            match Self::parse_operation(opcodes, pc) {
                Ok((op, new_pc)) => {
                    operations.push(op);
                    pc = new_pc;
                }
                Err(e) => {
                    operations.push(Operation::Invalid);
                    failed_opcodes.push(e);
                    pc += 1;
                }
            }
        }

        (operations, failed_opcodes)
    }

    fn parse_operation(
        bytecode: &[u8],
        mut pc: usize,
    ) -> Result<(Operation, usize), OpcodeParseError> {
        let opcode = Opcode::try_from(bytecode[pc])?;
        let op = Operation::from(opcode, bytecode, &mut pc)?;
        pc += 1;
        Ok((op, pc))
    }

    fn calculate_code_size(operations: &[Operation]) -> u32 {
        operations
            .iter()
            .map(|op| match op {
                Operation::Push((size, _)) => (size + 1) as u32,
                _ => 1,
            })
            .sum()
    }
}

impl From<Vec<Operation>> for Program {
    fn from(operations: Vec<Operation>) -> Self {
        let code_size = Self::calculate_code_size(&operations);
        Self {
            operations,
            code_size,
        }
    }
}

/// Returns the number of input and output stack elements of the given opcode.
pub const fn stack_io(op: &Operation) -> (u8, u8) {
    match op {
        Operation::Stop => (0, 0),
        Operation::Add => (2, 1),
        Operation::Mul => (2, 1),
        Operation::Sub => (2, 1),
        Operation::Div => (2, 1),
        Operation::Sdiv => (2, 1),
        Operation::Mod => (2, 1),
        Operation::SMod => (2, 1),
        Operation::Addmod => (3, 1),
        Operation::Mulmod => (3, 1),
        Operation::Exp => (2, 1),
        Operation::SignExtend => (2, 1),
        Operation::Lt => (2, 1),
        Operation::Gt => (2, 1),
        Operation::Slt => (2, 1),
        Operation::Sgt => (2, 1),
        Operation::Eq => (2, 1),
        Operation::IsZero => (1, 1),
        Operation::And => (2, 1),
        Operation::Or => (2, 1),
        Operation::Xor => (2, 1),
        Operation::Not => (1, 1),
        Operation::Byte => (2, 1),
        Operation::Shl => (2, 1),
        Operation::Shr => (2, 1),
        Operation::Sar => (2, 1),
        Operation::Keccak256 => (2, 1),
        Operation::Address => (0, 1),
        Operation::Balance => (1, 1),
        Operation::Origin => (0, 1),
        Operation::Caller => (0, 1),
        Operation::Callvalue => (0, 1),
        Operation::CalldataLoad => (1, 1),
        Operation::CallDataSize => (0, 1),
        Operation::CallDataCopy => (3, 0),
        Operation::Codesize => (0, 1),
        Operation::Codecopy => (3, 0),
        Operation::Gasprice => (0, 1),
        Operation::ExtcodeCopy => (4, 0),
        Operation::ReturnDataSize => (0, 1),
        Operation::ReturnDataCopy => (3, 0),
        Operation::ReturnDataLoad => (1, 1),
        Operation::ExtcodeHash => (1, 1),
        Operation::BlockHash => (1, 1),
        Operation::ExtcodeSize => (1, 1),
        Operation::Coinbase => (0, 1),
        Operation::Timestamp => (0, 1),
        Operation::Number => (0, 1),
        Operation::Prevrandao => (0, 1),
        Operation::Gaslimit => (0, 1),
        Operation::Chainid => (0, 1),
        Operation::SelfBalance => (0, 1),
        Operation::Basefee => (0, 1),
        Operation::BlobHash => (1, 1),
        Operation::BlobBaseFee => (0, 1),
        Operation::Pop => (1, 0),
        Operation::Mload => (1, 1),
        Operation::Mstore => (2, 0),
        Operation::Mstore8 => (2, 0),
        Operation::Sload => (1, 1),
        Operation::Sstore => (2, 0),
        Operation::Jump => (1, 0),
        Operation::Jumpi => (2, 0),
        Operation::Msize => (0, 1),
        Operation::Gas => (0, 1),
        Operation::Tload => (1, 1),
        Operation::Tstore => (2, 0),
        Operation::Mcopy => (3, 0),
        Operation::Push0 => (0, 1),
        Operation::Create => (3, 1),
        Operation::Call => (7, 1),
        Operation::CallCode => (7, 1),
        Operation::Return => (2, 0),
        Operation::DelegateCall => (6, 1),
        Operation::Create2 => (4, 1),
        Operation::StaticCall => (6, 1),
        Operation::Revert => (2, 0),
        Operation::Invalid => (0, 0),
        Operation::SelfDestruct => (1, 0),
        Operation::PC { .. } => (0, 1),
        Operation::Jumpdest { .. } => (0, 0),
        Operation::Push(_) => (0, 1),
        Operation::Dup(_) => (0, 1),
        Operation::Swap(n) => (*n + 1, *n + 1),
        Operation::Log(n) => (*n + 2, 0),
    }
}
