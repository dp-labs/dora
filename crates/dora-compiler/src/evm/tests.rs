use crate::context::Context;
use crate::evm::program::{Operation, Program};
use crate::evm::{CompileOptions, EVMCompiler};
use crate::Compiler;
use dora_primitives::Bytecode;
use num_bigint::BigUint;

macro_rules! assert_snapshot {
    ($operations:expr) => {
        assert_snapshot!($operations, false)
    };
    ($operations:expr, $is_eof:expr) => {
        let program = Program {
            operations: $operations,
            code_size: 0,
            is_eof: $is_eof,
        };
        let context = Context::new();
        let compiler = EVMCompiler::new(&context);
        // Compile EVM Bytecode to MLIR EVM Dialect
        let module = compiler
            .compile(&program, &(), &CompileOptions::default())
            .expect("failed to compile program");
        insta::assert_snapshot!(module.module().as_operation());
    };
}

#[test]
fn push_push_add() {
    let (a, b) = (BigUint::from(11_u8), BigUint::from(31_u8));
    let operations = vec![
        Operation::Push((1_u8, a.clone())),
        Operation::Push((1_u8, b.clone())),
        Operation::Add,
    ];
    assert_snapshot!(operations);
}

#[test]
fn push_push_sub() {
    let (a, b) = (BigUint::from(11_u8), BigUint::from(5_u8));
    let operations = vec![
        Operation::Push((1_u8, a.clone())),
        Operation::Push((1_u8, b.clone())),
        Operation::Sub,
    ];
    assert_snapshot!(operations);
}

#[test]
fn push_push_mul_div() {
    let (a, b) = (BigUint::from(1_u8), BigUint::from(2_u8));
    let operations = vec![
        Operation::Push((1_u8, a.clone())),
        Operation::Push((1_u8, b.clone())),
        Operation::Mul,
        Operation::Push((1_u8, b.clone())),
        Operation::Div,
    ];
    assert_snapshot!(operations);
}

#[test]
fn push0_pop() {
    let operations = vec![Operation::Push0, Operation::Pop];
    assert_snapshot!(operations);
}

#[test]
fn push0_iszero_pop() {
    let operations = vec![Operation::Push0, Operation::IsZero, Operation::Pop];
    assert_snapshot!(operations);
}

#[test]
fn program_from_eof_bytecode() {
    let bytecode = hex_literal::hex!("ef0001010004020001013c04006d00008000056080806040526004361015e100035f80fd5f3560e01c9081633fb5c1cb14e100e081638381f58a14e1009c5063d09de08a14e100045fe0ffd534e100875f7ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffc360112e1005c5f547fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff8114e100086001015f555f80f37f4e487b71000000000000000000000000000000000000000000000000000000005f52601160045260245ffd5f80fd5f80fd34e100335f7ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffc360112e100086020905f548152f35f80fd5f80fd34e1003460207ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffc360112e100086004355f555f80f35f80fd5f80fda3646970667358221220cc6570f0f9fb641c08c7d9d1c36810bd19987855bcd8f9ccde7cfcd8670b41fa6c6578706572696d656e74616cf564736f6c63782c302e382e32372d646576656c6f702e323032342e372e32342b636f6d6d69742e64353139363430342e6d6f64006b");
    let bytecode = Bytecode::new_raw(bytecode.into());
    let program: Program = bytecode.into();
    insta::assert_snapshot!(format!("{:#?}", program.operations));
}
