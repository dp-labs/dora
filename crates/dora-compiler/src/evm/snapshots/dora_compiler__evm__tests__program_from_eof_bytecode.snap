---
source: crates/dora-compiler/src/evm/tests.rs
expression: "format!(\"{:#?}\", program.operations)"
snapshot_kind: text
---
[
    Push(
        (
            1,
            128,
        ),
    ),
    Dup(
        1,
    ),
    Push(
        (
            1,
            64,
        ),
    ),
    MStore,
    Push(
        (
            1,
            4,
        ),
    ),
    CalldataSize,
    Lt,
    IsZero,
    RJumpI(
        3,
    ),
    Push0,
    Dup(
        1,
    ),
    Revert,
    Push0,
    CalldataLoad,
    Push(
        (
            1,
            224,
        ),
    ),
    Shr,
    Swap(
        1,
    ),
    Dup(
        2,
    ),
    Push(
        (
            4,
            1068876235,
        ),
    ),
    Eq,
    RJumpI(
        224,
    ),
    Dup(
        2,
    ),
    Push(
        (
            4,
            2206332298,
        ),
    ),
    Eq,
    RJumpI(
        156,
    ),
    Pop,
    Push(
        (
            4,
            3500007562,
        ),
    ),
    Eq,
    RJumpI(
        4,
    ),
    Push0,
    RJump(
        65493,
    ),
    CallValue,
    RJumpI(
        135,
    ),
    Push0,
    Push(
        (
            32,
            115792089237316195423570985008687907853269984665640564039457584007913129639932,
        ),
    ),
    CalldataSize,
    Add,
    Slt,
    RJumpI(
        92,
    ),
    Push0,
    SLoad,
    Push(
        (
            32,
            115792089237316195423570985008687907853269984665640564039457584007913129639935,
        ),
    ),
    Dup(
        2,
    ),
    Eq,
    RJumpI(
        8,
    ),
    Push(
        (
            1,
            1,
        ),
    ),
    Add,
    Push0,
    SStore,
    Push0,
    Dup(
        1,
    ),
    Return,
    Push(
        (
            32,
            35408467139433450592217433187231851964531694900788300625387963629091585785856,
        ),
    ),
    Push0,
    MStore,
    Push(
        (
            1,
            17,
        ),
    ),
    Push(
        (
            1,
            4,
        ),
    ),
    MStore,
    Push(
        (
            1,
            36,
        ),
    ),
    Push0,
    Revert,
    Push0,
    Dup(
        1,
    ),
    Revert,
    Push0,
    Dup(
        1,
    ),
    Revert,
    CallValue,
    RJumpI(
        51,
    ),
    Push0,
    Push(
        (
            32,
            115792089237316195423570985008687907853269984665640564039457584007913129639932,
        ),
    ),
    CalldataSize,
    Add,
    Slt,
    RJumpI(
        8,
    ),
    Push(
        (
            1,
            32,
        ),
    ),
    Swap(
        1,
    ),
    Push0,
    SLoad,
    Dup(
        2,
    ),
    MStore,
    Return,
    Push0,
    Dup(
        1,
    ),
    Revert,
    Push0,
    Dup(
        1,
    ),
    Revert,
    CallValue,
    RJumpI(
        52,
    ),
    Push(
        (
            1,
            32,
        ),
    ),
    Push(
        (
            32,
            115792089237316195423570985008687907853269984665640564039457584007913129639932,
        ),
    ),
    CalldataSize,
    Add,
    Slt,
    RJumpI(
        8,
    ),
    Push(
        (
            1,
            4,
        ),
    ),
    CalldataLoad,
    Push0,
    SStore,
    Push0,
    Dup(
        1,
    ),
    Return,
    Push0,
    Dup(
        1,
    ),
    Revert,
    Push0,
    Dup(
        1,
    ),
    Revert,
]
